newDeref: possible false-positive with generics

First of all, thank you for the awesome linter, I'm in love with go-critic ❤️

It seems like [newDeref](https://go-critic.com/overview.html#newderef) reports a false-positive when type parameters are used:

```go
func test[T any]() T {
	return *new(T)
}
```

> newDeref: replace `*new(T)` with `T(nil)`

Which is not possible because the conversion is invalid:

> cannot convert nil to T ​compiler: InvalidConversion

Found [here](https://github.com/junk1tm/assert/blob/main/assert.go#L105).

golangci-lint version: 1.49.0
gocritic version: v0.6.5 (installed with brew)

Seems like the following code when compiled via `go tool compile -S main.go` does not produce any `runtime.newobject` calls 🤔 

```go
func main() {
	_ = test[time.Time]()
}

//go:noinline
func test[T fmt.Stringer]() string {
	s := (*new(T)).String()
	return s
}

```

@quasilyte does it mean that the optimization you mentioned actually exists? (I'm a compiler newbie 😅)

`go tool compile -m main.go` reports that `main.go:14:12: new() does not escape`

--- a/checkers/newDeref_checker.go
+++ b/checkers/newDeref_checker.go
@@ -7,6 +7,7 @@ import (
 	"github.com/go-critic/go-critic/checkers/internal/lintutil"
 	"github.com/go-critic/go-critic/framework/linter"
 	"github.com/go-toolsmith/astcast"
+	"golang.org/x/exp/typeparams"
 	"golang.org/x/tools/go/ast/astutil"
 )
 
@@ -33,6 +34,10 @@ func (c *newDerefChecker) VisitExpr(expr ast.Expr) {
 	call := astcast.ToCallExpr(deref.X)
 	if astcast.ToIdent(call.Fun).Name == "new" {
 		typ := c.ctx.TypeOf(call.Args[0])
+		// allow *new(T) if T is a type parameter, see #1272 for details
+		if typeparams.IsTypeParam(typ) {
+			return
+		}
 		zv := lintutil.ZeroValueOf(astutil.Unparen(call.Args[0]), typ)
 		if zv != nil {
 			c.warn(expr, zv)

--- a/checkers/testdata/newDeref/negative_tests118.go
+++ b/checkers/testdata/newDeref/negative_tests118.go
@@ -1,0 +1,8 @@
+//go:build go1.18
+// +build go1.18
+
+package checker_test
+
+func genericNew[T any]() T {
+	return *new(T)
+}

